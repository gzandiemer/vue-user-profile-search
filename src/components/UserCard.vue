<template>
  <div>
    <router-link :to="`/user/${user.login.uuid}`">
      <img
        :src="user.picture.thumbnail"
        alt="User picture"
      >
      <h2>{{ formattedName }}</h2>
      <p>{{ user.email }}</p>
    </router-link>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
export default defineComponent({
  props: {
    user: {
      type: Object,
      required: true
    }
  },
  computed: {
    formattedName() {
      const first =
        this.user.name.first.charAt(0).toUpperCase() +
        this.user.name.first.slice(1);
      const last =
        this.user.name.last.charAt(0).toUpperCase() +
        this.user.name.last.slice(1);
      return `${first} ${last}`;
    }
  }
});
</script>

<style></style>
