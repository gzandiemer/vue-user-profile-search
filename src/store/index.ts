import { createStore } from 'vuex';
import { User } from '../types';

interface State {
  users: User[];
  page: number;
}

export default createStore({
  state: {
    users: [],
    page: 1,
  } as State,
  mutations: {
    setUsers(state: State, users: User[]) {
      state.users = state.users.concat(users);
    },
    incrementPage(state: State){
      state.page ++;
    }
  },
  actions: {
    async fetchUsers({ commit, state }) {
      const response = await fetch(`https://randomuser.me/api/?results=25&page=${state.page}`);
      const data = await response.json();
      commit('setUsers', data.results);
      commit('incrementPage');
    }
  },
  getters: {
    getUser: (state: State) => (id: string) => {
      return state.users.find((user) => user.login.uuid === id);
    }
  }
});

