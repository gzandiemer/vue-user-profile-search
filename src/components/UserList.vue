<template>
  <div>
    <user-card
      v-for="user in filteredUsers"
      :key="user.login.uuid"
      :user="user"
    />
    <button @click="fetchUsers">
      More results...
    </button>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, onMounted, onUnmounted } from 'vue';
import { useStore } from 'vuex';
import UserCard from '../components/UserCard.vue';
import { User } from '../types';

export default defineComponent({
  name: 'UserList',
  components: { UserCard },
  props: {
    search: String,
    gender: String
  },
  setup(props) {
    const store = useStore();

    const fetchUsers = async () => {
      await store.dispatch('fetchUsers');
    };

    const users = computed(() => store.state.users as User[]);

    const filteredUsers = computed(() => {
      return users.value.filter((user: User) => {
        const searchString = props.search ? props.search.toLowerCase() : '';
        const nameMatch = `${user.name.first} ${user.name.last}`
          .toLowerCase()
          .includes(searchString);
        const genderMatch = props.gender === '' || user.gender === props.gender;
        return nameMatch && genderMatch;
      });
    });
    onMounted(async () => {
      // Add scroll event listener on component mount
      window.addEventListener('scroll', handleScroll);
      // Fetch initial users on component mount
      await fetchUsers();
    });

    onUnmounted(() => {
      // Remove scroll event listener before component unmount
      window.removeEventListener('scroll', handleScroll);
    });

    function handleScroll() {
      // Check if user has scrolled to the bottom of the page
      if (window.innerHeight + window.scrollY >= document.body.offsetHeight) {
        // Load more users
        fetchUsers();
      }
    }

    return { fetchUsers, users, filteredUsers };
  }
});
</script>

<style scoped></style>
