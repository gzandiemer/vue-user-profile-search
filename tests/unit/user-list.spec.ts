import { mount } from "@vue/test-utils";
import UserList from "@/components/UserList.vue";
import UserCard from "@/components/UserCard.vue";
import { createStore } from "vuex";
import { cloneDeep } from 'lodash';
import { State } from "@/store";

interface MyStoreOptions {
  state: State;
  actions: any;
  mutations: any;
  getters: any;
}

describe('UserList.vue', () => {
  let state: State;
  let actions: any;
  let mutations: any;
  let getters: any;
  let localStore: any;
  let localStoreConfig: MyStoreOptions;
  let wrapper: any;

  const mockFetchUsers = jest.fn();

  const usersData = [{
    gender: 'male',
    name: {
      title: 'Mr',
      first: 'John',
      last: 'Doe'
    },
    location: {
      street: {
        number: 123,
        name: 'Main St'
      },
      city: 'Anytown',
      state: 'AnyState',
      country: 'AnyCountry',
      postcode: '12345',
      coordinates: {
        latitude: '50.0000',
        longitude: '50.0000'
      },
      timezone: {
        offset: '+1:00',
        description: 'AnyTimezone'
      }
    },
    email: 'john.doe@example.com',
    login: {
      uuid: '1a2b3c',
      username: 'johndoe',
      password: 'password',
      salt: 'salt',
      md5: 'md5',
      sha1: 'sha1',
      sha256: 'sha256'
    },
    phone: '123-456-7890',
    picture: {
      large: 'https://randomuser.me/api/portraits/men/1.jpg',
      medium: 'https://randomuser.me/api/portraits/med/men/1.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/men/1.jpg'
    },
    dob: {
      date: '1970-01-01T00:00:00Z',
      age: 50
    },
    registered: {
      date: '2022-01-01T00:00:00Z',
      age: 1
    },
    cell: '123-456-7890',
    id: {
      name: 'idname',
      value: 'idvalue'
    },
      nat: 'US'
    }, {
      gender: 'female',
      name: {
        title: 'Ms',
        first: 'Jane',
        last: 'Doe'
      },
      location: {
        street: {
          number: 100,
          name: 'Anystreet'
        },
        city: 'Anytown',
        state: 'AnyState',
        country: 'AnyCountry',
        postcode: '12345',
        coordinates: {
          latitude: '50.0000',
          longitude: '50.0000'
        },
        timezone: {
          offset: '+1:00',
          description: 'AnyTimezone'
        }
      },
      email: 'john.doe@example.com',
      login: {
        uuid: '9z8y7w',
        username: 'johndoe',
        password: 'password',
        salt: 'salt',
        md5: 'md5',
        sha1: 'sha1',
        sha256: 'sha256'
      },
      phone: '123-456-7890',
      picture: {
        large: 'https://randomuser.me/api/portraits/men/1.jpg',
        medium: 'https://randomuser.me/api/portraits/med/men/1.jpg',
        thumbnail: 'https://randomuser.me/api/portraits/thumb/men/1.jpg'
      },
      dob: {
        date: '1970-01-01T00:00:00Z',
        age: 50
      },
      registered: {
        date: '2022-01-01T00:00:00Z',
        age: 1
      },
      cell: '123-456-7890',
      id: {
        name: 'idname',
        value: 'idvalue'
      },
        nat: 'US'
      }];

      
      beforeEach(() => {
        state = {
          users: usersData,  // predefined usersData
          selectedUser: null,
          searchQuery: '',
          filter: '',
          page: 1,
          gender: ''
        };
    
        actions = {
          fetchUsers: mockFetchUsers
        };
    
        mutations = {
          setUsers: jest.fn(),
          setSelectedUser: jest.fn(),
          clearSelectedUser: jest.fn(),
          setSearchQuery: jest.fn(),
          setFilter: jest.fn(),
          setGender: jest.fn(),
          incrementPage: jest.fn()
        };
    
        getters = {
          getUser: jest.fn(),
        };
    
        localStoreConfig = {
          state,
          actions,
          mutations,
          getters
        };
    
        localStore = createStore(cloneDeep(localStoreConfig));
      });
    
      beforeEach(() => {
        wrapper = mount(UserList, { global: { plugins: [localStore] } });
      });
    
      it('should render correctly', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });
    
      it('dispatches "fetchUsers" on component mount', () => {
        expect(mockFetchUsers).toHaveBeenCalled();
      });
    
      it('renders a UserCard for each user in state', () => {
        const userCards = wrapper.findAllComponents(UserCard);
        expect(userCards.length).toBe(localStore.state.users.length);
      });
    
      it('emits "user-selected" when a UserCard emits "user-selected"', async () => {
        await wrapper.vm.$nextTick();
        const userCards = wrapper.findAllComponents(UserCard);
        if(userCards.length > 0) {
          const selectedUser = localStore.state.users[0];
          await userCards[0].vm.$emit('user-selected', selectedUser);
          expect(wrapper.emitted('user-selected')).toBeTruthy();
          expect(wrapper.emitted('user-selected')?.[0]).toEqual([selectedUser]);
        } else {
          throw new Error('No UserCard components found');
        }
      });
    });