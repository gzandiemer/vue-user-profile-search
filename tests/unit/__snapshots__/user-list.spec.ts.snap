// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList.vue should render correctly 1`] = `
<div class="user-list">
  <div class="user-card"><img class="user-picture" src="https://randomuser.me/api/portraits/thumb/men/1.jpg" alt="User picture">
    <div class="user-info">
      <h2>John Doe</h2>
      <p>john.doe@example.com</p>
    </div>
  </div>
  <div class="user-card"><img class="user-picture" src="https://randomuser.me/api/portraits/thumb/men/1.jpg" alt="User picture">
    <div class="user-info">
      <h2>Jane Doe</h2>
      <p>john.doe@example.com</p>
    </div>
  </div><button> More results... </button>
</div>
`;
